compiler/src/iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp

```diff
@@ -203,6 +206,15 @@ summarizeDispatchWorkgroupsOp(DispatchWorkgroupsOp regionOp) {
   int64_t bestEstimatedCost = kMinEstimatedCost;
   regionOp.getWorkgroupBody().walk([&](Operation *op) {
     TypeSwitch<Operation *>(op)
+        .Case<linalg::SoftmaxOp>([&](auto op) {
+          int64_t estimatedCost = estimateLinalgExtOpCost(op);
+          if (estimatedCost < bestEstimatedCost)
+            return;
+          bestEstimatedCost = estimatedCost;
+          bestOp = op;
+          LLVM_DEBUG(llvm::dbgs() << "// new best op: '" << bestOp->getName()
+                                  << "', cost: " << bestEstimatedCost << "\n");
+        })
```

Comment:
THis change itself is fine. Just move this to a method `estimateLinalgSoftmaxOp` and this is good.

> Linalg :: Yield, Index and Softmax don't percolate down through the `isa<linalg::LinalgOp>` check.
> These ops, as per the upstream llvm-project, are added as ones [that do not correspond to library calls](https://github.com/llvm/llvm-project/blob/f5cb9cb59d7c9c6ac3d5c41c677f68c9b75d34a3/mlir/include/mlir/Dialect/Linalg/IR/LinalgOps.td#L26).

The confusing part here is that `LinalgOp` in the `.td` file is not same as the `LinalgOp` in C++. The one in `.td` file is actually scoped to just the `.td` file and has no meaning outside of it (the naming is unfortunate). The `LinalgOp` in C++ is all operations that implement the `LinalgStructuredOpInterface` here https://github.com/llvm/llvm-project/blob/740ee00a4c1d841204f57c88c04fa62e38bd1322/mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.td#L182 . Right there you see the source of confusion 
```
def LinalgStructuredInterface
    : OpInterface<"LinalgOp", [DestinationStyleOpInterface]> {
```
In the `.td` file it is called `LinalgStructuredInterface` but the `OpInterface<"LinalgOp>` means the generated header files use `LinalgOp` as the name of the interface operation in C++. Thats an unfortunate choice.

Effectively all Linalg named ops are just a `linalg.generic` with a fixed `indexing_maps`, `iterator_types`, and `region`.  So they all implement the `LinalgStructuredOpInterface` (in the `.td` file which translates to implementing the `LinalgOp` interface in C++). 

(FYI @nicolasvasilache / @qcolombet / @matthias-springer , this is a bit confusing conflation of names that needs to be resolved I think... Its hard to explain, let alone understand)

`SoftMax` on the other hand is NOT a `linalg.generic`, it is an aggregate op. So this distinction makes sense. It is NOT specified to implement the `LinalgStructuredInterface` in the `.td` definition. So it does NOT implement the C++ `LinalgOp` interface. So the `cast` fails.

Question though is why it even is getting through to the `cast`... The check above here https://github.com/openxla/iree/blob/2b9ade513428ed9c0714134f120a50024c5718b5/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp#L651 should have avoided even getting here to the `cast`. Not sure why that happened.

Reply:
> Question though is why it even is getting through to the `cast`... The check above here
> 
> https://github.com/openxla/iree/blob/2b9ade513428ed9c0714134f120a50024c5718b5/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp#L651
> 
> should have avoided even getting here to the `cast`. Not sure why that happened.

My bad! Apologies.

The correct place where it crashes is [Upstream LinalgOp::Softmax -> reifyResultShapes](https://github.com/llvm/llvm-project/blob/7675f541f75baa20e8ec007cd625a837e89fc01f/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L2390).

The issue, though, is same => `LinalgOp::Softmax` not percolating through `cast<linalg::LinalgOp>` because of NOT implementing [LinalgStructuredInterface](https://github.com/llvm/llvm-project/blob/740ee00a4c1d841204f57c88c04fa62e38bd1322/mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.td#L182).

Something to perhaps address upstream?
A possible solution being => Get rid of the `cast` and have a Softmax specific implementation for `reifyResultShapes` upstream? Because the input/output shape in case of Softmax is SAME, it should be a trivial implementation?
We didn't face this issue in case of `LinalgExt::Softmax` because the `cast` was on `LinalgExt` and with the interface's method's implementation in, it worked.

Reply:
> > Question though is why it even is getting through to the `cast`... The check above here
> > https://github.com/openxla/iree/blob/2b9ade513428ed9c0714134f120a50024c5718b5/compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp#L651
> > 
> > should have avoided even getting here to the `cast`. Not sure why that happened.
> 
> My bad! Apologies.
> 
> The correct place where it crashes is [Upstream LinalgOp::Softmax -> reifyResultShapes](https://github.com/llvm/llvm-project/blob/7675f541f75baa20e8ec007cd625a837e89fc01f/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L2390).
> 
> The issue, though, is same => `LinalgOp::Softmax` not percolating through `cast<linalg::LinalgOp>` because of NOT implementing [LinalgStructuredInterface](https://github.com/llvm/llvm-project/blob/740ee00a4c1d841204f57c88c04fa62e38bd1322/mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.td#L182).
>

> Something to perhaps address upstream? A possible solution being => Get rid of the `cast` and have a Softmax specific implementation for `reifyResultShapes` upstream? Because the input/output shape in case of Softmax is SAME, it should be a trivial implementation? We didn't face this issue in case of `LinalgExt::Softmax` because the `cast` was on `LinalgExt` and with the interface's method's implementation in, it worked.

Yeah, thats the fix. That simple change should make it work. Thanks for traiging!

Reply:
I've added a patch for review upstream : https://github.com/llvm/llvm-project/pull/67790